File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/ErrorStatementSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class ErrorStatementSyntax : ExecutableStatementSyntax
	{
		internal readonly KeywordSyntax _errorKeyword;

		internal readonly ExpressionSyntax _errorNumber;

		internal static Func<ObjectReader, object> CreateInstance;

		internal KeywordSyntax ErrorKeyword => _errorKeyword;

		internal ExpressionSyntax ErrorNumber => _errorNumber;

		internal ErrorStatementSyntax(SyntaxKind kind, KeywordSyntax errorKeyword, ExpressionSyntax errorNumber)
			: base(kind)
		{
			base._slotCount = 2;
			AdjustFlagsAndWidth(errorKeyword);
			_errorKeyword = errorKeyword;
			AdjustFlagsAndWidth(errorNumber);
			_errorNumber = errorNumber;
		}

		internal ErrorStatementSyntax(SyntaxKind kind, KeywordSyntax errorKeyword, ExpressionSyntax errorNumber, ISyntaxFactoryContext context)
			: base(kind)
		{
			base._slotCount = 2;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(errorKeyword);
			_errorKeyword = errorKeyword;
			AdjustFlagsAndWidth(errorNumber);
			_errorNumber = errorNumber;
		}

		internal ErrorStatementSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax errorKeyword, ExpressionSyntax errorNumber)
			: base(kind, errors, annotations)
		{
			base._slotCount = 2;
			AdjustFlagsAndWidth(errorKeyword);
			_errorKeyword = errorKeyword;
			AdjustFlagsAndWidth(errorNumber);
			_errorNumber = errorNumber;
		}

		internal ErrorStatementSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 2;
			KeywordSyntax keywordSyntax = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax != null)
			{
				AdjustFlagsAndWidth(keywordSyntax);
				_errorKeyword = keywordSyntax;
			}
			ExpressionSyntax expressionSyntax = (ExpressionSyntax)reader.ReadValue();
			if (expressionSyntax != null)
			{
				AdjustFlagsAndWidth(expressionSyntax);
				_errorNumber = expressionSyntax;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_errorKeyword);
			writer.WriteValue(_errorNumber);
		}

		static ErrorStatementSyntax()
		{
			CreateInstance = (ObjectReader o) => new ErrorStatementSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(ErrorStatementSyntax), (ObjectReader r) => new ErrorStatementSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.ErrorStatementSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _errorKeyword, 
				1 => _errorNumber, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new ErrorStatementSyntax(base.Kind, newErrors, GetAnnotations(), _errorKeyword, _errorNumber);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new ErrorStatementSyntax(base.Kind, GetDiagnostics(), annotations, _errorKeyword, _errorNumber);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitErrorStatement(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/ErrorStatementSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class ErrorStatementSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExecutableStatementSyntax
		Friend ReadOnly _errorKeyword As KeywordSyntax

		Friend ReadOnly _errorNumber As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property ErrorKeyword As KeywordSyntax
			Get
				Return Me._errorKeyword
			End Get
		End Property

		Friend ReadOnly Property ErrorNumber As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax
			Get
				Return Me._errorNumber
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errorKeyword As KeywordSyntax, ByVal errorNumber As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)
			MyBase.New(kind)
			MyBase._slotCount = 2
			MyBase.AdjustFlagsAndWidth(errorKeyword)
			Me._errorKeyword = errorKeyword
			MyBase.AdjustFlagsAndWidth(errorNumber)
			Me._errorNumber = errorNumber
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errorKeyword As KeywordSyntax, ByVal errorNumber As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind)
			MyBase._slotCount = 2
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(errorKeyword)
			Me._errorKeyword = errorKeyword
			MyBase.AdjustFlagsAndWidth(errorNumber)
			Me._errorNumber = errorNumber
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal errorKeyword As KeywordSyntax, ByVal errorNumber As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)
			MyBase.New(kind, errors, annotations)
			MyBase._slotCount = 2
			MyBase.AdjustFlagsAndWidth(errorKeyword)
			Me._errorKeyword = errorKeyword
			MyBase.AdjustFlagsAndWidth(errorNumber)
			Me._errorNumber = errorNumber
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 2
			Dim keywordSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax)
				Me._errorKeyword = keywordSyntax
			End If
			Dim expressionSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)
			If (expressionSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(expressionSyntax)
				Me._errorNumber = expressionSyntax
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitErrorStatement(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.ErrorStatementSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Dim num As Integer = i
			If (num = 0) Then
				greenNode = Me._errorKeyword
			ElseIf (num = 1) Then
				greenNode = Me._errorNumber
			Else
				greenNode = Nothing
			End If
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._errorKeyword, Me._errorNumber)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ErrorStatementSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._errorKeyword, Me._errorNumber)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._errorKeyword, IObjectWritable))
			writer.WriteValue(DirectCast(Me._errorNumber, IObjectWritable))
		End Sub
	End Class
End Namespace